Project.toml

```diff
@@ -26,6 +26,7 @@ SpecialFunctions = "276daf66-3868-5448-9aa4-cd146d93841b"
 Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
 StatsFuns = "4c63d2b9-4356-54db-8cca-17b64c39e42c"
 Tracker = "9f7883ad-71c0-57eb-9f7f-b5c9e6d3789c"
+Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"
```

Comment:
It seems only ZygoteRules is needed (if `Zygote.@nograd Bijectors._debug` is moved to Bijectors)

Reply:
We may need Zygote because it has the definition of the `pullback` method that we are using here.

Reply:
No, `pullback` is actually defined in ZygoteRules.

Reply:
https://github.com/FluxML/ZygoteRules.jl/blob/b0dc38ed8b857f0beb79afe165cb853e16a1a261/src/adjoint.jl#L21

Reply:
BTW due to the same reason probably Zygote (currently) does not have to be an (optional) dependency of AdvancedHMC (https://github.com/TuringLang/AdvancedHMC.jl/pull/161/files#r366268742).

Reply:
Yes but I think calling it will error unless `Zygote` is loaded because the 2-arg method is defined in `Zygote`. Since `DistributionsAD` loads `Zygote` anyways, the method will be loaded. But I think it it's better to load `Zygote` here as well in case `DistributionsAD` stops loading `Zygote` in the future.

Reply:
Hmm I think I disagree. It seems the main purpose of ZygoteRules is for other packages to be able to define custom adjoints and refer to `pullback` without having to depend on Zygote and all its dependencies (I think ForwardDiff2 uses a similar approach with ChainRules). Especially if a package wants to support different AD backends, I think it is nice if users are not forced to install every single AD backend and all its dependencies.

It also seems that (currently) ForwardDiff is the default AD backend, and hence a user has to actively choose Zygote in which case, I guess, users should be expected to load Zygote explicitly. Nevertheless, I guess it might be useful to check that Zygote is actually loaded if the backend is switched to Zygote and provide a helpful error message in case it's not.

Reply:
@devmotion is exactly right regarding the motivation for `ZygoteRules`.

It would be a really great thing if the entire Turing stack depended only on `ZygoteRules`, rather than `Zygote`, precisely to ensure that users that don't want to install Zygote and its dependencies.

This will become even more acute once we've moved Zygote onto ChainRules (see [this issue](https://github.com/FluxML/Zygote.jl/pull/366)).

Reply:
Supporting Zygote optionally sounds good to me if that's the path we want to take. But note that we currently support Tracker by default too, and Zygote is currently used to define MvNormal adjoints for Tracker use. So if we want to optionally support Zygote, we will have to optionally support Tracker as well, or break Tracker and MvNormal. Both sound like bad options to me. Alternatively, we can support all these AD packages by default which currently won't increase our loading time at all because we load Zygote anyways in DistributionsAD to support Tracker. I think this is a better short-term solution.

Reply:
I'm not in favour of making all AD backends optional right away and breaking Tracker and MvNormal :stuck_out_tongue:

IMO it's fine to have the explicit dependency in DistributionsAD at the moment, even though I think it would be a nice long-term goal to get rid of such heavy dependencies. But I see that it might be difficult to achieve this (or even impossible if it's an indirect dependency from some non-Turing package).

Still, even if DistributionsAD depends on Turing, I think Turing shouldn't add Zygote as a new explicit dependency. Depending on ZygoteRules seems to be sufficient and the intended approach. Additionally, not depending on Zygote seems like moving into the right direction if one wants to pursue the long-term goal of slimming down dependencies.

Reply:
Ok that can be done. But when Zygote is no longer loaded in Turing by default, that will be a breaking change so we need to be mindful of that. I will replace Zygote with ZygoteRules.

Reply:
I'm not sure what's the best way to ensure that users can't use `ZygoteAD` if Zygote is not loaded, but maybe one could something similar to https://github.com/TuringLang/AdvancedHMC.jl/pull/161, i.e., use Requires and define optional backends such as `ZygoteAD` only if Zygote is loaded. More concretely, one could define
```julia
ADBackend(::Val) = error("The requested AD backend is not available. Make sure to load all required packages.")

function gradient_logp(
    θ::AbstractVector{<:Real},
    vi::VarInfo,
    model::Model,
    sampler::Sampler
)
    return gradient_logp(θ, vi, model, sampler, getADtype(sampler))
end

# optional dependencies
function __init__()
    @require Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f" begin 
         struct ZygoteAD <: ADBackend end

         ADBackend(::Val{:zygote}) = ZygoteAD

         function gradient_logp(
            θ::AbstractVector{<:Real},
            vi::VarInfo,
            model::Model,
            sampler::Sampler,
            ::ZygoteAD
        )
            T = typeof(getlogp(vi))

            # Specify objective function.
            function f(θ)
                new_vi = VarInfo(vi, sampler, θ)
                return getlogp(runmodel!(model, new_vi, sampler))
            end

            # Compute forward and reverse passes.
            l, ȳ = Zygote.pullback(f, θ)
            ∂l∂θ = ȳ(1)[1]
    
            return l, ∂l∂θ
        end
    end
end
```

Reply:
Yes that's reasonable.

Reply:
One could even update the default backend (I edited the comment above) and it does require neither Zygote nor ZygoteRules as dependencies :smiley: